libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,821 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/param.h>
+#include <sys/ktrace.h>
+#include <sys/mman.h>
+#include <sys/sysctl.h>
+
+#include <stdlib.h>
+#include <stdatomic.h>
+
+#include "debug.h"
+#include "rtld.h"
+#include "rtld_c18n.h"
+#include "rtld_libc.h"
+#include "rtld_utrace.h"
+
+/*
+ * Sealers for RTLD privileged information
+ */
+uintptr_t sealer_pltgot, sealer_jmpbuf, sealer_tramp;
+
+/* Use utrace() to log compartmentalisation-related events */
+const char *ld_compartment_utrace;
+
+/* Enable compartmentalisation */
+const char *ld_compartment_enable;
+
+/* Simulate overhead during compartment transitions */
+const char *ld_compartment_overhead;
+
+/* Read the .c18n.signature ELF section */
+const char *ld_compartment_sig;
+
+/*
+ * libthr support
+ */
+void (*thr_thread_start)(struct pthread *);
+
+void
+_rtld_thread_start_init(void (*p)(struct pthread *))
+{
+	assert((cheri_getperm(p) & CHERI_PERM_EXECUTIVE) == 0);
+	assert(thr_thread_start == NULL);
+	thr_thread_start = tramp_intern(NULL, &(struct tramp_data) {
+		.target = p,
+		.defobj = obj_from_addr(p),
+		.sig = (struct func_sig) {
+			.valid = true,
+			.reg_args = 1, .mem_args = false, .ret_args = NONE
+		}
+	});
+}
+
+void (*thr_sighandler)(int, siginfo_t *, void *);
+
+void
+_rtld_sighandler_init(void (*p)(int, siginfo_t *, void *))
+{
+	assert((cheri_getperm(p) & CHERI_PERM_EXECUTIVE) == 0);
+	assert(thr_sighandler == NULL);
+	thr_sighandler = tramp_intern(NULL, &(struct tramp_data) {
+		.target = p,
+		.defobj = obj_from_addr(p),
+		.sig = (struct func_sig) {
+			.valid = true,
+			.reg_args = 3, .mem_args = false, .ret_args = NONE
+		}
+	});
+}
+
+/*
+ * Policies
+ */
+typedef ssize_t string_handle;
```

Comment:
Why handle? A handle is an actual thing, not information about a thing like size/capacity? And why is this then signed if it's for size/capacity, are you actually using -1 for something?

Reply:
It's an index into a flat buffer of strings, and -1 means string not found.